### ISSUE TITLE
Update to repository workflow document

### ISSUE THREAD

**Issue by gold2718** [ESMCI/cime#972]
@ESMCI/developers :
I have updated the [CIME Repository Workflow](https://github.com/ESMCI/cime/wiki/Repository-workflow) document to clarify how issues should be created and assigned before commencing work and how 'Reviewers' and 'Assignees' are to be used in pull requests.
These changes are meant to help facilitate workflow and avoid team members duplicating work due to a lack of information.
Please provide feedback in this issue (which I will leave open for a week or so) and/or notify us here if you make an edit.

**Comment by billsacks** [comment]
@gold2718 - thank you very much for this! There is some good stuff here. I especially appreciate the description of reviewers vs. assignee - I was wondering about this.

1.

> Before working on a new feature, or a bug fix, make sure there is an issue related to the work and assign yourself to it.

While I'm fine with this suggestion, this does not seem to be the current practice: I see a lot of PRs that don't have any issues associated with them. Are we planning to require at least one issue for each PR? If not, this should be removed.

2.

> Make sure to start from the most recent version of master that passed the testing

Is there a way to determine what the most recent version of master is that passed the testing? We had some conversations around this a few months ago, but I still don't see a way to see this. It would be nice (I might say _critical_) to have this ability if we don't have a separate next branch.

There is also some ambiguity here around what is meant by "the testing". If I'm doing developments that can be tested purely within CIME, then I can generally get away with using a bleeding-edge version of master. But if I'm doing developments that require testing with active CESM components, then I usually end up regretting starting from a bleeding-edge version, and find things go more smoothly when I make a branch off of the version used in the last beta tag (i.e., "the testing" = "CESM beta testing"). Should that be noted explicitly here?

3.

> Bug fixes should start from the earliest commit that introduced the bug

As far as I can tell, this has not been the practice so far. Are cime developers going to commit to this? It seems to have some value, but also creates extra work, in that you need to track down the original commit that had the bug. I'm willing to commit to it if others are, but if others are unwilling, then this shouldn't be part of the official documentation.

4. 

Is there guidance as to whether branches should be created on the ESMCI fork or your personal fork?

5.

Particularly if branches are to be created on the ESMCI fork, then is there a branch naming convention?

I see:

> git branch username/component/bug-fix commit-with-bug

> git branch username/component/todo maint-cimeX.Y

which suggests a branch naming convention, but nowhere does this document explicitly say we should be using a branch naming convention. In fact, the above seems like it might just be taken from the ACME branch naming convention, which doesn't necessarily apply here (particularly the mention of 'component').

6.

Do I understand correctly that you do NOT need to explicitly select the Assignee as a Reviewer? If so, I'd suggest rewording this:

> When making a pull request, assign as 'Reviewers' everyone you think should review the code. Assign an 'Assignee' to manage the pull request. 

to something like:

When making a pull request, assign a single 'Assignee' to manage the pull request. If you would like others besides the assignee to review the code, then assign them as 'Reviewers'.

**Comment by jgfouca** [comment]
@billsacks I'm working on (2) now.

**Comment by billsacks** [comment]
Thanks, @jgfouca ! I realize that this requires some non-trivial effort, so I appreciate it!

**Comment by jgfouca** [comment]
I'm going to go with a hacky solution that is at least easy.

**Comment by billsacks** [comment]
Fine by me.

**Comment by gold2718** [comment]
Breaking this up to facilitate discussion:

> 
> > Before working on a new feature, or a bug fix, make sure there is an issue related to the work and assign yourself to it.
> 
> While I'm fine with this suggestion, this does not seem to be the current practice: I see a lot of PRs that don't have any issues associated with them. Are we planning to require at least one issue for each PR? If not, this should be removed.
>
This is a suggestion for a new requirement motivated by duplicated work which was due to a lack of communication. I feel this would help us keep track of what is being worked on in CIME. If anyone has a better suggestion, I would love to see it.

**Comment by billsacks** [comment]
>>> Before working on a new feature, or a bug fix, make sure there is an issue related to the work and assign yourself to it.

>> While I'm fine with this suggestion, this does not seem to be the current practice: I see a lot of PRs that don't have any issues associated with them. Are we planning to require at least one issue for each PR? If not, this should be removed.

> This is a suggestion for a new requirement motivated by duplicated work which was due to a lack of communication. I feel this would help us keep track of what is being worked on in CIME. If anyone has a better suggestion, I would love to see it.

I'm happy with this suggestion, I just wanted to know that there is buy-in from @jedwards4b , @mvertens , @jgfouca and other frequent committers.

**Comment by gold2718** [comment]
> Do I understand correctly that you do NOT need to explicitly select the Assignee as a Reviewer? If so, I'd suggest rewording this:
>
I would be sad if we had many pull requests without review (we have too many of those now). However, if the consensus is that we don't need review, your wording is okay with me.
I'm a little waffley (technical term) about the single Assignee because someone could be added as an Assignee if they need to work on the PR. Not sure of a great wording for this.

**Comment by billsacks** [comment]
>> Do I understand correctly that you do NOT need to explicitly select the Assignee as a Reviewer? If so, I'd suggest rewording this:

> I would be sad if we had many pull requests without review (we have too many of those now). However, if the consensus is that we don't need review, your wording is okay with me.
> I'm a little waffley (technical term) about the single Assignee because someone could be added as an Assignee if they need to work on the PR. Not sure of a great wording for this.

This is kind of my source of confusion about Assignees vs. Reviewers. In my mind, the Assignee is implicitly also a reviewer (so doesn't need to be explicitly assigned the capital-R Reviewer role), but I'm not sure if that's what others think. It sounds like your thinking is that the Assignee should also be explicitly listed as a Reviewer (capital R). I'm fine with that; I just want guidance on what's expected here.

**Comment by billsacks** [comment]
>>> Do I understand correctly that you do NOT need to explicitly select the Assignee as a Reviewer? If so, I'd suggest rewording this:

>> I would be sad if we had many pull requests without review (we have too many of those now). However, if the consensus is that we don't need review, your wording is okay with me.
>> I'm a little waffley (technical term) about the single Assignee because someone could be added as an Assignee if they need to work on the PR. Not sure of a great wording for this.

> This is kind of my source of confusion about Assignees vs. Reviewers. In my mind, the Assignee is implicitly also a reviewer (so doesn't need to be explicitly assigned the capital-R Reviewer role), but I'm not sure if that's what others think. It sounds like your thinking is that the Assignee should also be explicitly listed as a Reviewer (capital R). I'm fine with that; I just want guidance on what's expected here.

To further add to my confusion, we have the "Code review" section of the PR template. I'm not sure how that interacts with these other roles, if at all. Should we remove that?

**Comment by jedwards4b** [comment]
I don't see anything particularly troubling here - i will try to conform.

**Comment by billsacks** [comment]
>> Make sure to start from the most recent version of master that passed the testing

> Is there a way to determine what the most recent version of master is that passed the testing? We had some conversations around this a few months ago, but I still don't see a way to see this. It would be nice (I might say critical) to have this ability if we don't have a separate next branch.

I just edited the documentation to describe how to find the most recently tested version. If the most recently tested version passed, then you're in luck. If not, I'm not sure what to suggest.

**Comment by jgfouca** [comment]
@billsacks the Back button on the dashboard. You can keep clicking it until you see all green.

**Comment by billsacks** [comment]
@jgfouca - thanks. I have updated the documentation with these instructions.

**Comment by rljacob** [comment]
This all looks good.  Its important that issue be opened as soon as one realizes there's a thing that needs to be done, before a branch is made or code is written.  That's hard because most of us like to try writing code first.
